api/src/main/java/bio/terra/tanagra/underlay/entitygroup/CriteriaOccurrence.java

```diff
@@ -0,0 +1,199 @@
+import bio.terra.tanagra.query.Query;
+import bio.terra.tanagra.query.TableVariable;
+import bio.terra.tanagra.serialization.UFEntityGroup;
+import bio.terra.tanagra.underlay.AuxiliaryData;
+import bio.terra.tanagra.underlay.AuxiliaryDataMapping;
+import bio.terra.tanagra.underlay.DataPointer;
+import bio.terra.tanagra.underlay.Entity;
+import bio.terra.tanagra.underlay.EntityGroup;
+import bio.terra.tanagra.underlay.EntityGroupMapping;
+import bio.terra.tanagra.underlay.Relationship;
+import bio.terra.tanagra.underlay.RelationshipMapping;
+import com.google.common.collect.ImmutableMap;
+import java.util.List;
+import java.util.Map;
+
+public class CriteriaOccurrence extends EntityGroup {
+  private static final String CRITERIA_ENTITY_NAME = "criteria";
+  private static final String OCCURRENCE_ENTITY_NAME = "occurrence";
+  private static final String OCCURRENCE_TO_CRITERIA_RELATIONSHIP_NAME = "occurrenceToCriteria";
+  private static final String OCCURRENCE_TO_PRIMARY_RELATIONSHIP_NAME = "occurrenceToPrimary";
+
+  private static final String CRITERIA_PRIMARY_ROLLUP_COUNT_AUXILIARY_DATA_NAME =
+      "criteriaPrimaryRollupCount";
+  private static final AuxiliaryData CRITERIA_PRIMARY_ROLLUP_COUNT_AUXILIARY_DATA =
+      new AuxiliaryData(
+          CRITERIA_PRIMARY_ROLLUP_COUNT_AUXILIARY_DATA_NAME, List.of("criteriaId", "primaryCount"));
+
+  private final Entity criteriaEntity;
+  private final Entity occurrenceEntity;
+  private final Entity primaryEntity;
+
+  private CriteriaOccurrence(Builder builder) {
+    super(builder);
+    this.criteriaEntity = builder.criteriaEntity;
+    this.occurrenceEntity = builder.occurrenceEntity;
+    this.primaryEntity = builder.primaryEntity;
+  }
+
+  public static CriteriaOccurrence fromSerialized(
+      UFEntityGroup serialized,
+      Map<String, DataPointer> dataPointers,
+      Map<String, Entity> entities,
+      String primaryEntityName) {
+    Entity criteriaEntity = deserializeEntity(serialized, CRITERIA_ENTITY_NAME, entities);
+    Entity occurrenceEntity = deserializeEntity(serialized, OCCURRENCE_ENTITY_NAME, entities);
+    Entity primaryEntity = entities.get(primaryEntityName);
+
+    Map<String, Relationship> relationships =
+        Map.of(
+            OCCURRENCE_TO_CRITERIA_RELATIONSHIP_NAME,
+                new Relationship(
+                    OCCURRENCE_TO_CRITERIA_RELATIONSHIP_NAME, occurrenceEntity, criteriaEntity),
+            OCCURRENCE_TO_PRIMARY_RELATIONSHIP_NAME,
+                new Relationship(
+                    OCCURRENCE_TO_PRIMARY_RELATIONSHIP_NAME, occurrenceEntity, primaryEntity));
+    Map<String, AuxiliaryData> auxiliaryData =
+        Map.of(
+            CRITERIA_PRIMARY_ROLLUP_COUNT_AUXILIARY_DATA_NAME,
+            CRITERIA_PRIMARY_ROLLUP_COUNT_AUXILIARY_DATA);
+
+    EntityGroupMapping sourceDataMapping =
+        EntityGroupMapping.fromSerializedForSourceData(
+            serialized.getSourceDataMapping(), dataPointers, relationships, auxiliaryData);
```

Comment:
I don't understand auxiliary data. Why is it specified in a generic fashion like this? How do the "criteriaId" and "primaryCount" strings work? Are they generated attributes that can be requested?

Reply:
- Auxiliary data is basically tables other than the relationship table that we need to execute queries. e.g. static rollup count tables.
- It's specified in a generic fashion so that we don't need to modify the serialization classes when we add a new entity group. I don't know if this is right approach long-term -- an alternative that I actually implemented before changing to this approach, is to have each entity group define its own serialization format. So I'm not wedded to this approach, and think we should revisit it when we add the next entity group.
- "criteriaId" and "primaryCount" strings are just aliases that the indexing workflows expect in the input SQL queries. Similar to the comment below, I think the indexing workflows should be changed to not expect specific column names, but I didn't want to tackle that in this already very big PR.